scripts/hypervisor-facts.py

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env python3
+# This script generates hypervisor (xen) facts out of the hostvars of each VM
+import json
+import optparse
+import re
+
+
+def main():
+  p = optparse.OptionParser(
+    description='Sets the relevant ansible facts needed for a xen hypervisor.',
+    usage='%prog [hypervisor hostname] [path to file containing hostvars in json]')
+  options, arguments = p.parse_args()
+  if len(arguments) == 2:
+    hypervisor_host = arguments[0]
+    original_facts = json.loads(open(arguments[1]).read())
+    facts = generateFacts(original_facts, hypervisor_host)
+    print(json.dumps(facts))
+  else:
+    p.print_help()
+
+
+def generateFacts(original_facts, hypervisor_host):
+  # facts are the hostvars of the current hypervisor host
+  facts = original_facts[hypervisor_host] if hypervisor_host in original_facts else {}
+  cidr_suffix = facts['vm_facts_default_cidr_suffix'] if 'vm_facts_default_cidr_suffix' in facts else '/24'
+  # List of hostnames that have missing VM vars. This lists will be printed in tasks for easier debugging
+  failed_names = {'interfaces': [], 'description': []}
+
+  # Create list if it doesn't already exist
+  if 'xen_vman_vms' not in facts:
+    facts['xen_vman_vms'] = []
+
+  # Traverse every host defined in hostvars
+  for host in original_facts.keys():
+    # Ignore physical hosts
+    if 'vm' not in original_facts[host]:
+      continue
+
+    # config is the vm dict of a specific VM host
+    config = original_facts[host]['vm']
+    config['name'] = host
+    # Ignore manually defined vms
+    if any((d['name'] == host and 'org' in config and d['org'] == config['org']) for d in facts['xen_vman_vms']):
+      continue
+
+    # Set precise connection type for xen, depending on general storage_type
+    if 'storage_type' in config and config['storage_type'] == 'blockdevice':
+      config['storage_type'] = 'iscsi'
+    else:
+      config['storage_type'] = 'nfs'
```

Comment:
This will set the storage type to nfs if not specified, but it isn't required to set the storage_type for xen_vman. There is an option to specify a default strorage_type `xen_vman_default_storage_type`

Reply:
@Mr-Pi But there is no default option in our storage role, because the decision to use `filesystem`/`zvol` results in the insertion in different lists which can not easily be defaulted.
If I remove this setting and `xen_vman_default_storage_type` is set to `blockdevice`, the storage will create a filesystem but `xen_vman` expects an iSCSI blockdevice.